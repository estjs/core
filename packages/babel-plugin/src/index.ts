import { transformProgram } from './program';
import { replaceProps } from './transformers/props';
import { transformJSX } from './jsx';
import type { PluginObj } from '@babel/core';

export default function (): PluginObj {
  return {
    name: '@estjs/babel-plugin',

    manipulateOptions({ filename }, parserOpts) {
      if (filename.endsWith('.ts') || filename.endsWith('.tsx')) {
        parserOpts.plugins.push('typescript');
      }
      parserOpts.plugins.push('jsx');
    },

    visitor: {
      Program: transformProgram,
      FunctionDeclaration: replaceProps,
      ArrowFunctionExpression: replaceProps,
      JSXElement: transformJSX,
      JSXFragment: transformJSX,
    },
  };
}
