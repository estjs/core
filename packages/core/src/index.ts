export * from './signal';
export * from './template';
export * from './version';
export * from './server';

export type * from '../types/index.d.ts';
