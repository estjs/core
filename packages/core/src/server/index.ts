import { isFunction, isObject, startsWith } from 'essor-shared';
import type { EssorNode } from '../../types';

interface TemplateMap {
  [key: number]: {
    template: string;
    prop?: Record<string, any>;
  };
}

type Props = Record<string, any>;

export function ssrtmpl(templates: string[] = []): TemplateMap {
  return templates.reduce((acc, template, index) => {
    acc[index + 1] = { template };
    return acc;
  }, {} as TemplateMap);
}

function jsonToAttrs(json: Record<string, any>): string {
  return Object.entries(json)
    .map(([key, value]) => `${key}=${JSON.stringify(value)}`)
    .join(' ');
}

export function ssr(
  template: TemplateMap | EssorNode | ((props: Props) => string),
  props: Props,
): string {
  if (isFunction(template)) {
    return template(props);
  }

  const childrenMap: Record<string, EssorNode[]> = {};
  const newTemplate: TemplateMap = {};
  if (isObject(template)) {
    Object.entries(template).forEach(([key, tmpl]) => {
      const prop = props[key];
      if (prop) {
        Object.keys(prop).forEach(propKey => {
          if (startsWith(propKey, 'on') && isFunction(prop[propKey])) {
            delete prop[propKey];
          }
        });

        if (prop.children) {
          prop.children.forEach(([child, idx]: [any, number]) => {
            childrenMap[idx] = [...(childrenMap[idx] || []), child];
          });
          delete prop.children;
        }
      }

      newTemplate[key] = { template: tmpl.template, prop };
    });
  }

  return Object.entries(newTemplate)
    .map(([key, { template: tmpl, prop }]) => {
      let str = tmpl;
      if (prop) {
        str += ` ${jsonToAttrs(prop)}`;
      }
      if (childrenMap[key]) {
        str += childrenMap[key].map(child => ssr(child, prop)).join('');
      }

      return str;
    })
    .join('');
}

export function renderToString(component: (...args: any[]) => string, props: Props): string {
  return ssr(component, props);
}

export function hydrate(
  component: { mount: (root: HTMLElement) => void },
  root: HTMLElement,
): void {
  root.innerHTML = '';
  component.mount(root);
}
