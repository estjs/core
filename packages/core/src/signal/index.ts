export { useSignal, useEffect, useComputed, signalObject, isSignal, isComputed } from './signal';
export { type Signal, type Computed } from './signal';
export { createStore, StoreActions } from './store';
