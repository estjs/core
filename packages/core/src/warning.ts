export function warn(msg: string, ..._args: any[]): void;
export function warn(msg: string, ...args): void {
  // eslint-disable-next-line prefer-spread
  console.warn.apply(console, [`[Essor warn]: ${msg}`].concat(args) as [string, ...any[]]);
}
