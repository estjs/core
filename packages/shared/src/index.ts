export {
  noop,
  _toString,
  extend,
  hasChanged,
  coerceArray,
  hasOwn,
  startsWith,
  escape,
  isExclude,
  ExcludeType,
  generateUniqueId,
  isBrowser,
} from './comm';
export {
  isString,
  isObject,
  isArray,
  isMap,
  isSet,
  isWeakMap,
  isWeakSet,
  isFunction,
  isNil,
  isPromise,
  isSymbol,
  isFalsy,
  isPlainObject,
  isPrimitive,
  isHTMLElement,
  isStringNumber,
  StringNumber,
} from './is';
export { camelCase, kebabCase, capitalizeFirstLetter } from './name';
export { warn, info, error } from './console';
