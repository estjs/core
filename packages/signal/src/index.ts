export {
  useSignal,
  useEffect,
  useComputed,
  isSignal,
  isComputed,
  useReactive,
  isReactive,
  shallowReactive,
  shallowSignal,
  clearReactive,
  signalObject,
  useBatch,
  toRaw,
} from './signal';
export { useWatch } from './watch';
export { type Signal, type Computed } from './signal';
export { createStore, StoreActions } from './store';
