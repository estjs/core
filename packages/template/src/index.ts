export { TemplateNode } from './template-node';
export { ComponentNode } from './component-node';
export { h, template, Fragment, isJsxElement } from './template';
export { nextTick } from './utils';
export { onMount, onDestroy, useInject, useProvide, InjectionKey } from './hooks';

export { renderToString, renderSSG } from './server';

export type * from '../types/index.d.ts';
