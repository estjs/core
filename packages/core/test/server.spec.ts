import { hydrate, renderToString, ssr, ssrtmpl } from '../src';

describe('service', () => {
  let App;

  beforeEach(() => {
    const _tmpl$ = ssrtmpl(['<div>', 'Hello, ', '!', '</div>']);

    function AppFn(props) {
      const name = 'John';

      return ssr(_tmpl$, {
        '1': {
          id: props.id,
          class: props.class,
          children: [[() => name, 2]],
        },
      });
    }
    App = AppFn;
  });

  afterEach(() => {
    App = null;
  });
  it('should return an empty object for empty templates array', () => {
    expect(ssrtmpl()).toEqual({});
  });

  it('should create a template map with correct indices', () => {
    const templates = ['template1', 'template2', 'template3'];
    const result = ssrtmpl(templates);
    expect(result).toEqual({
      1: { template: 'template1' },
      2: { template: 'template2' },
      3: { template: 'template3' },
    });
  });

  it('should render a template with props', () => {
    const props = {
      id: 'test',
      class: 'test-class',
    };
    const result = ssr(App, props);
    expect(result).toMatchInlineSnapshot(
      `"[object Object] id=""test"" class=""test-class""Hello, John!</div>"`,
    );
  });

  it('should render a component to a string', () => {
    const props = {
      template1: {
        id: 'test',
        class: 'test-class',
      },
    };
    const result = renderToString(App, props);
    expect(result).toMatchInlineSnapshot(
      `"[object Object] id="undefined" class="undefined"Hello, John!</div>"`,
    );
  });
  it('should mount component to the root element', () => {
    document.body.innerHTML = '<div id="root"></div>';
    const component = {
      mount: vitest.fn(),
    };
    hydrate(component, document.querySelector('#root')!);
    expect(component.mount).toHaveBeenCalled();
  });
});
