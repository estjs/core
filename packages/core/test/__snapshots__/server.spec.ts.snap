// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`service > should render a template with props > Hello, <div id="test" className="test-class"> 1`] = `"Hello <div>,!,</div> id="test" className="test-class" chidren="() => 123,1""`;

exports[`service > should render a template with props 1`] = `"Hello <div>,!,</div> id="test" className="test-class" chidren="() => 123,1""`;
